File: bench/fnv1a64.mjs
// BOT-ASSISTED

const CAP = 65536
const RUNS = 2048
const OFFSET = 0xcbf29ce484222325n
const PRIME_LO = 0x1b3
const WANT = 0xb0a1ea8560222325n
const buf = new Uint8Array(CAP)
const pat = Buffer.from("0123456789abcdef")

for (let ind = 0; ind < CAP; ind++) buf[ind] = pat[ind & 15]

function fnv1a64(hash, src) {
  let hi = Number((hash >> 32n) & 0xffffffffn)
  let lo = Number(hash & 0xffffffffn)
  for (let ind = 0; ind < src.length; ind++) {
    lo = (lo ^ src[ind]) >>> 0
    const carry = (Math.imul(lo >>> 16, PRIME_LO) + (Math.imul(lo & 0xffff, PRIME_LO) >>> 16)) >>> 16
    hi = (Math.imul(hi, PRIME_LO) + carry + ((lo << 8) >>> 0)) >>> 0
    lo = Math.imul(lo, PRIME_LO) >>> 0
  }
  return (BigInt(hi) << 32n) | BigInt(lo)
}

function fmt(hash) {
  return `0x${hash.toString(16).padStart(16, "0")}`
}

let hash = OFFSET
for (let rep = 0; rep < RUNS; rep++) hash = fnv1a64(hash, buf)

if (hash !== WANT) {
  throw Error(`mismatch: expected ${fmt(WANT)}; got ${fmt(hash)}`)
}

File: bench/fnv1a64.lua
-- BOT-ASSISTED

local ffi = require("ffi")
local bit = require("bit")

local CAP = 65536
local RUNS = 2048
local OFFSET = 0xcbf29ce484222325ULL
local PRIME = 0x100000001b3ULL
local WANT = 0xb0a1ea8560222325ULL
local buf = ffi.new("uint8_t[?]", CAP)
local pat = "0123456789abcdef"
local int64 = ffi.typeof("int64_t")

for ind = 0, CAP - 1 do
  buf[ind] = pat:byte((ind % 16) + 1)
end

local function fnv1a64(hash, src, len)
  for ind = 0, len - 1 do
    local low = tonumber(hash % 256ULL)
    hash = (hash + int64(bit.bxor(low, src[ind]) - low)) * PRIME
  end
  return hash
end

local function fmt(hash)
  local high = tonumber(hash / 0x100000000ULL)
  local low = tonumber(hash % 0x100000000ULL)
  return string.format("0x%08x%08x", high, low)
end

local hash = OFFSET
for _ = 1, RUNS do
  hash = fnv1a64(hash, buf, CAP)
end
if hash ~= WANT then
  error("mismatch: expected " .. fmt(WANT) .. "; got " .. fmt(hash))
end
